staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -471,38 +458,79 @@ func (s preparedGenericAPIServer) Run(stopCh <-chan struct{}) error {
 		// - once drained, http Server Shutdown is invoked with a timeout of 2s,
 		//   net/http waits for 1s for the peer to respond to a GO_AWAY frame, so
 		//   we should wait for a minimum of 2s
-		delayedStopOrDrainedCh = drainedCh.Signaled()
 		shutdownTimeout = 2 * time.Second
 		klog.V(1).InfoS("[graceful-termination] using HTTP Server shutdown timeout", "ShutdownTimeout", shutdownTimeout)
 	}
 
-	// pre-shutdown hooks need to finish before we stop the http server
-	preShutdownHooksHasStoppedCh, stopHttpServerCh := make(chan struct{}), make(chan struct{})
+	notAcceptingNewRequestCh := s.lifecycleSignals.NotAcceptingNewRequest
+	drainedCh := s.lifecycleSignals.InFlightRequestsDrained
+	stopHttpServerCh := make(chan struct{})
 	go func() {
 		defer close(stopHttpServerCh)
 
-		<-delayedStopOrDrainedCh
-		<-preShutdownHooksHasStoppedCh
+		notAcceptingNewRequestOrDrainedCh := notAcceptingNewRequestCh.Signaled()
+		if s.ShutdownSendRetryAfter {
+			notAcceptingNewRequestOrDrainedCh = drainedCh.Signaled()
+		}
+
+		<-notAcceptingNewRequestOrDrainedCh
 	}()
 
+	// Use an stop channel to allow graceful shutdown without dropping audit events
```

Comment:
I went ahead and I split from my PR the commit that is documenting the current dependencies within shutdown sequence: https://github.com/kubernetes/kubernetes/pull/110081

In order to make it easier to review I suggest to:
1) First merge my PR (documenting current state)
2) Split from this PR that are more of refactoring changes that doesn't really change behavior, e.g.:
(a) moving AuditBackend run our of NonBlockRunning (that still leave it as pre-shutdown)
(b) introduction of preshutdownhook signal (that is still not blocking anything)
[In fact (1) and (2) can be made in parallel]

3) Once the above changes, the core changes in this PR will get smaller (thus a bit easier to review), so
- update this PR to also update the diagram introduced by my PR
- review that with those changes

@tkashem @aojea - thoughts?

Reply:
> moving AuditBackend run our of NonBlockRunning (that still leave it as pre-shutdown)

Or actually even moving it our of pre-shutdown already in the first PR - to reduce the size of individual changes.

Reply:
- https://github.com/kubernetes/kubernetes/pull/110094
- https://github.com/kubernetes/kubernetes/pull/110096